src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java

```diff
@@ -47,51 +76,75 @@ public int getChunkZ() {
         return cz;
     }
 
+    private boolean isInvalid(int x, int y, int z) {
+        if (y < 0 || y >= worldHeight) return true;
+        return ((x | z) & ~0xF) != 0;
+    }
+
     @Override
     public boolean isTrue(int x, int y, int z) {
-        return store[x][z][y];
+        if (isInvalid(x, y, z)) return false;
+        int status = data.get((y * PAGE_SIZE) + (x >> 1));
+        return ((status >>> (z + (((x & 1) == 1) ? 16 : 0))) & 1) != 0;
     }
 
-    @Override
-    public void setTrue(int x, int y, int z) {
-        if (y >= store[0][0].length || y < 0)
+    private void set(int x, int y, int z, boolean val) {
+        if (isInvalid(x, y, z))
             return;
-        store[x][z][y] = true;
+        int point = (y * PAGE_SIZE) + (x >> 1);
+        int bit = z + (((x & 1) == 1) ? 16 : 0);
+        int source;
+        do {
+            source = data.get(point);
+            int target = source;
+            if (val) {
+                target |= 1 << bit;
+            } else {
+                target &= ~(1 << bit);
+            }
+            if (data.compareAndSet(point, source, target)) break;
+        } while (true);
```

Comment:
Please include comments inside this method explaining the operations you are taking, especially the bit shifting.